Add --cfg and --rustc-cfg flags to output compiler configuration

volks73 opened the pull request:
This PR is my attempt to address #8923.

I have added the `--cfg` flag to the `cargo rustc` subcommand and the `--rustc-cfg` flag to the `cargo build`, `cargo check`, `cargo test`, and `cargo bench` subcommands, respectively. Both versions of the flag, `--cfg` and `--rustc-cfg`, do the same thing, but I thought it looked weird for the `cargo rustc` subcommand to be `cargo rustc --rustc-cfg`. The following example invocations are possible, once stabilized:

- `cargo rustc --cfg`
- `cargo build --rustc-cfg`
- `cargo check --rustc-cfg`
- `cargo test --rustc-cfg`
- `cargo bench --rustc-cfg`

In each case, compilation is aborted and only compiler configuration is emitted. All of the context creation and configuration is still executed, but execution of the compilation job is aborted. Similar to the `--unit-graph` implementation, the `--cfg/--rustc-cfg` flag is hidden, marked as a "unstable", and requires the `-Z unstable-options` flag at the moment. A complete example invocation with this PR would be:

```bash
$ cargo +nightly rustc -Z unstable-options --cfg
```

I am open to alternatives for the flag name. I have thought of `--compiler-cfg`, `--compile-cfg`, and `--target-cfg`, but I went with `--rustc-cfg` because it is the Rust compiler (rustc) configuration (cfg). The `--target-cfg` could be confusing because there are Cargo targets and Compiler targets. A lone `--cfg` for the build, check, test, and bench subcommands would also be ambiguous and configuration that is being displayed.

Originally, I was only going to add the `--cfg` flag to the `cargo rustc` subcommand, but build, check, test, and bench all have the `--unit-graph` flag, and I used that flag's implementation and existence as a template for this implementation. I am not opposed to having just the `--cfg` flag for the `cargo rustc` subcommand as originally proposed in #8923.

I discovered during my initial investigation to implement the feature and familiarization with the Cargo codebase, that as part of the build context creation and compilation process, Cargo internally calls the `rustc --print cfg` command for all targets and stores the output in the `RustcTargetData` type. It does this for the host and any explicitly defined targets for cross-compilation. Compilation features, such as `+crt-static`, added to the compilation process through Cargo environment variables or TOML configuration directives are added to the internal `rustc --print cfg` command call. So, the `--cfg/--rustc-cfg` just emits the contents of the `RustcTagetData` type as JSON. There is no need to call the `rustc --print cfg` command again. The implementation then becomes nearly identical to the `--unit-graph` implementation. 

The output is only in JSON, similar to the `--unit-graph` feature, instead of the key-value and name style of the `rustc --print cfg` output because it easily resolves issues related to multi-targets, explicit target (`--target <TRIPLE>`) versus host configurations, and cross compilation. Subcommands and other projects can parse the JSON to recreate the key-value and name style if desired based on a specific target or the host. Here is an example of the JSON output (formatted for humans, the actual output is condensed), which is also available as a comment in the `cargo::core::compiler::rustc_cfg` module:

```javascript
{
     "version": 1,
     "host": {
         "names": ["windows", "debug_assertions"],
         "arch":"x86_64",
         "endian":"little",
         "env":"msvc",
         "family":"windows",
         "features":["fxsr","sse","sse2"],
         "os":"windows",
         "pointer_width":"64",
         "vendor":"pc"
     },
     "targets": {
         "x86_64-unknown-linux-gnu": {
             "names": ["debug_assertions", "unix"],
             "arch":"x86_64",
             "endian":"little",
             "env":"gnu",
             "family":"unix",
             "features": ["fxsr","sse","sse2"],
             "os":"linux",
             "pointer_width":"64",
             "vendor":"unknown"
         },
         "i686-pc-windows-msvc": {
             "names": ["windows", "debug_assertions"],
             "arch":"x86",
             "endian":"little",
             "env":"msvc",
             "family":"windows",
             "features":["fxsr","sse","sse2"],
             "os":"windows",
             "pointer_width":"32",
             "vendor":"pc"
         }
     }
 }
```

I decided to remove the "target_" prefix from the relevant configurations to reduce "noise" in the output. Again, I am open to alternatives or suggestions on the JSON format.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @Eh2406 (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/cargo/blob/master/CONTRIBUTING.md) for more information.

ehuss commented:
Thanks for the PR!  Just a heads up, with the holidays, it may take a while to get a proper look at this.

Just my initial reaction, I'm a little uneasy about adding so many flags.  AFAIK, the cfg values only depend on the `--target` flag, is that correct?  The reason `--unit-graph` was done that way is because it depends on the command and pretty much all the flags, so it couldn't really work as a separate command.

I would also avoid hard-coding keys and stripping prefixes, and just dump the contents of the Cfg map.  Is there a particular reason to not display the whole thing?

volks73 commented:
>Thanks for the PR! Just a heads up, with the holidays, it may take a while to get a proper look at this.

Your welcome, and I understand. Happy holidays!

>Just my initial reaction, I'm a little uneasy about adding so many flags. AFAIK, the cfg values only depend on the --target flag, is >that correct? The reason --unit-graph was done that way is because it depends on the command and pretty much all the flags, so >it couldn't really work as a separate command.

Yes, I believe the cfg values are only dependent on the `--target` option(s) and the reasoning for the `--unit-graph` flag makes sense. The `--cfg/--rustc-cfg` flag should probably only be added for the `cargo rustc` subcommand. Obtaining the compiler configuration for a specific Cargo target, like for tests, could be obtained with `cargo rustc --test --cfg` instead of `cargo test --rustc-cfg`. My original thought was to only add the flag to the `cargo rustc` subcommand anyways.  I will hold off on making this change until further confirmation is received.

>I would also avoid hard-coding keys and stripping prefixes, and just dump the contents of the Cfg map. Is there a particular reason to not display the whole thing?

Up until commit [234089](https://github.com/rust-lang/cargo/pull/9002/commits/234089ffbe73e666e1985c93f11d64c4db7ffd1a), the initial implementation was just a dump of the contents:

```javascript
{
    "version": 1,
    "host": {
        "names": ["windows", "debug_assertions"],
        "key_pairs": [
            {"target_arch": "x86_64"},
            {"target_endian": "little"},
            {"target_env": "msvc"},
            {"target_family": "windows"},
            {"target_feature": "fxsr"},
            {"target_feature": "sse"},
            {"target_feature": "sse2"},
            {"target_os": "windows"},
            {"target_pointer_width": "64"},
            {"target_vendor": "pc"}
        ]
    },
    "targets": [
        "x86_64-unknown-linux-gnu": {
            "names": ["debug_assertions", "unix"],
            "key_pairs": [
                {"target_arch": "x86_64"},
                {"target_endian": "little"},
                {"target_env": "gnu"},
                {"target_family": "unix"},
                {"target_feature": "fxsr"},
                {"target_feature": "sse"},
                {"target_feature": "sse2"},
                {"target_os": "linux"},
                {"target_pointer_width": "64"},
                {"target_vendor": "unknown"}
            ]
        },
         "i686-pc-windows-msvc": {
            "names": ["windows", "debug_assertions"],
            "key_pairs": [
                {"target_arch":"x86"},
                {"target_endian":"little"},
                {"target_env":"msvc"},
                {"target_family":"windows"},
                {"target_feature":"fxsr"},
                {"target_feature":"sse"},
                {"target_feature":"sse2"},
                {"target_os":"windows"},
                {"target_pointer_width":"32"},
                {"target_vendor":"pc"}
         }
    ]
}
```

And looking at the output, I see a lot of "target_" being repeated. I use repeated prefixes as a "code smell" or indicator that something should be refactored/restructured. From the consumer side of the JSON output, like in a crate or third party subcommand, I would not have `Cfg::target_os`, `Cfg::target_family`, `Cfg::target_env`, etc. methods/fields, but `Cfg::os`, `Cfg::family`, `Cfg::env`, etc. or even `Cfg::Target::os`, `Cfg::Target::family`, `Cfg::Target::env`, etc. So, I thought the JSON should mirror this. 

I recognize hard-coding and stripping prefixes is maybe not the most flexible or future-proof implementation. All of the output I tested with the `rustc --print cfg --target <TRIPLE>` command had the hard-coded fields, i.e. "target_env", "target_os", "target_endian", etc., albeit an empty string for targets without the value. This seemed to indicate that the "target_*" configurations were not going away and would always be present, but it is possible new configurations (prefixed with `target_` or not) might be added or used and this is the concern? I thought the `Version` field would help in these situations.

I don't really have a strong reason for the proposed JSON format other than "look and feel", smaller number of bytes, and empirical evidence that certain fields are always available, which is not great because it is really for machine consumption, not human consumption, bandwidth/performance is not a concern, and I am not familiar enough with future compiler changes. Honestly, I struggled with the format and have mentally run around in circles on it. So, I decided I reached my limits and it was time to seek advice and review. I would greatly appreciate some direction, suggestions, and/or alternative organization of the output (when time and capacity permit).

An example of me mentally running around in a circle, a hybrid approach that condenses "target_*" configurations to "target" array field but maintains some extensibility because everything is an array where new configurations can just be added as new maps, but indicates that both the "names" and "target" configurations/fields are ultimately optional and may not exist:

```javascript
{
    "version": 1,
    "host": [
        {"names": ["windows", "debug_assertions"]},
        {"target": [
            {"arch": "x86_64"},
            {"endian": "little"},
            {"env": "msvc"},
            {"family": "windows"},
            {"features": ["fxsr", "sse", "sse2"]},
            {"os": "windows"},
            {"pointer_width": "64"},
            {"vendor": "pc"}
        ]}
    ],
    "targets": [
        "i686-pc-windows-msvc": [
            {"names": ["windows", "debug_assertions"]},
            {"target": [
                {"arch":"x86"},
                {"endian":"little"},
                {"env":"msvc"},
                {"family":"windows"},
                {"features":["fxsr","sse","sse2"]},
                {"os":"windows"},
                {"pointer_width":"32"},
                {"vendor":"pc"}
            ]}
        ]
    ]
}
```

alexcrichton commented:
I would agree with @ehuss that we probably don't want to munge the values and instead just report exactly what rustc is telling us. I think that a flag also may not be necessary for this, we can probably just add this json blob to the ones already emitted?

volks73 commented:
> I think that a flag also may not be necessary for this, we can probably just add this json blob to the ones already emitted?

I am not aware of other flags that output JSON and skip compiling/building. Are you referring to the `--unit-graph` flag? If yes, then I am good with making this implementation, too. Would it just be:

```javascript
{
    "rustc": ["windows", "debug_assertions", "target_arch='x86'", "target_endian='little'", ...]
}
```

somewhere within the `--unit-graph` JSON?

alexcrichton commented:
Oh I thought that this was just a normal json message coming from Cargo during build? We can add it to `--unit-graph` as well but I think it's fine for us to just add other blobs coming out during a build, they should have an identifier on them and implementations should be skipping identifiers they don't recognize.

volks73 commented:
>Oh I thought that this was just a normal json message coming from Cargo during build?

The goal was to get the compiler configurations with Cargo modifications, such as those defined in environment variables or Cargo configuration files, without building/compiling the project, similar to the `rustc --print cfg` command. I have not looked into the JSON messages that are emitted during the Cargo build process to see if the compiler configurations are part of the output, but I could see where this information would also be useful in that context as well.

I looked at the [`--unit-graph` output](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#unit-graph). There is the `platform` field, which is `null` if compilation is for the host and a target triple if it is for cross-compilation. A `cfg`, `configurations`, `rustc`, `compiler_cfg`, `compiler_configurations`, `rustc_configurations`, or other similarly named field could be added after the `platform` field:

```javascript
...
"platform": null,
"compiler_configurations":["debug_assertions","windows","target_arch='x86'","target_endian='little'", ...],
"mode": "build",
...
```

I would still like to consider a separate flag for the `cargo rustc --cfg` command and not include the configuration in the `--unit-graph` output for two reasons: 

1. The `--unit-graph` feature appears to be further along and closer to stabilization. I do not want this feature to slow down the stabilization of the `--unit-graph` feature as the `--unit-graph` feature appears to be more broadly desired. The compiler configuration field could be added after stabilization as a Version 2 of the `--unit-graph` output since I could see the configuration information for each unit being useful as well.
2. A `cargo rustc --cfg` command would be a nice analog to the `rustc --print cfg` command, which prints the configuration but skips compilation. The compiler configuration would be buried in the `--unit-graph` output. The `cargo rustc --cfg` command is more "semantically meaningful" to me, i.e. "show me the compiler configuration for the cargo compilation process".

alexcrichton commented:
Ah ok I understand now, thanks for helping me to clarify. I missed that this was a pre-compilation step rather than something learned during compilation.

To add to what @ehuss was mentioning earlier, this feels like the current implementation is a bit heavyweight relative to what it's doing. For example this is downloading the entire crate graph, which it probably doesn't need if it's just scraping config values. Unfortunately we don't really have a great analog for something similar to this and where it would go today, so I don't know where it would best go.

volks73 commented:
> To add to what @ehuss was mentioning earlier, this feels like the current implementation is a bit heavyweight relative to what it's doing. For example this is downloading the entire crate graph, which it probably doesn't need if it's just scraping config values. Unfortunately we don't really have a great analog for something similar to this and where it would go today, so I don't know where it would best go.

I agree that downloading the entire crate graph just for the compiler configuration is too much and the original implementation is then very heavyweight. Thus, I have made the following changes to the implementation:

1. Removed the `--rustc-cfg` flag from the check, test, and build subcommands, i.e. only `cargo rustc --print-cfg` now exists.
2. Changed the `--cfg` flag for the rustc subcommand to `--print-cfg` to be more descriptive of the function. It is printing the configuration, not configuring it or making any changes.
3. Created a lighter weight implementation that does not download the crate graph and simply obtains the Rustc target data for the workspace based on the Cargo and package configurations. A completely separate "pathway" from the `--unit-graph` implementation is used. This also results in smaller footprint within the codebase. The `ops::print_cfg` function could be moved to the `commands::rustc` module for an even smaller footprint since only the `cargo rustc` command (currently) uses it. However, I thought the function might be useful in the future for other subcommands.
4. Changed the JSON format to not modify the field names. An example output is in the comments of the `compile::rustc_cfg` module (this could also be moved to the `commands::rustc` module if desired) and below. I was not sure about the double/single quotes for the value strings of the key-pair fields. The `rustc --print cfg` uses double quotes for the value, but these would have to be escaped in JSON. I think we could also just drop the quotes in the JSON values.

```javascript
{
    "version": 1,
    "host": [
        "windows",
        "debug_assertions",
        "target_arch='x86_64'",
        "target_endian='little'",
        "target_env='msvc'",
        "target_family='windows'",
        "target_feature='fxsr'",
        "target_feature='sse'",
        "target_feature='sse2'",
        "target_os='windows'",
        "target_pointer_width='64'",
        "target_vendor='pc'"
    ],
    "targets": [
        {
            "x86_64-unknown-linux-gnu": [
                "unix",
                "debug_assertions",
                "target_arch='x86_64'",
                "target_endian='little'",
                "target_env='gnu'",
                "target_family='unix'",
                "target_feature='fxsr'",
                "target_feature='sse'",
                "target_feature='sse2'",
                "target_os='linux'",
                "target_pointer_width='64'",
                "target_vendor='unknown'"
            ]
        },
        {
            "i686-pc-windows-msvc": [
                "windows",
                "debug_assertions",
                "target_arch='x86'",
                "target_endian='little'",
                "target_env='msvc'",
                "target_family='windows'",
                "target_feature='fxsr'",
                "target_feature='sse'",
                "target_feature='sse2'",
                "target_os='windows'",
                "target_pointer_width='32'",
                "target_vendor='pc'"
            ]
        }
    }
}
```

I think these changes make for a lighter weight implementation and keeps things as untouched as possible. This only addresses a flag/functionality for outputting just the compiler configuration. It does not address or change the unit graph output. I believe adding the compiler configuration to the unit graph output would be a separate issue and feature independent of this functionality.

alexcrichton commented:
Personally I still feel that a feature like this feels really "tacked on" to Cargo. This doesn't seem like it really has all that much benefit over just running `rustc --cfg` yourself, and it may be better for Cargo to implement a way to learn about `RUSTFLAGS` configuration? 

This doesn't feel like a general feature of Cargo but rather a very specific feature request. For example the current implementation just ignores everything else passed to `cargo rustc` effectively other than `--print-cfg`.

I'm not really sure of the best way to solve and design this myself, however. We could maybe take a leaf from rustc's book and have something like `cargo print rustc-cfg` or maybe even `cargo config --print rustc-cfg` or I dunno. Something which is a general way of "print out Cargo's internal information" which we can extend in the future rather than "here's a very specific print flag on one command"

volks73 commented:
>...it may be better for Cargo to implement a way to learn about RUSTFLAGS configuration? 

Do you mean Cargo's overall configuration, not just the environment variable? Cargo passes configurations to the compiler from its `.cargo/config.toml` hierarchy (`rustflags` field), `RUSTFLAGS`, CLI (`--target`), and I think maybe `cfg` attributes in code, too?

>Personally I still feel that a feature like this feels really "tacked on" to Cargo. This doesn't seem like it really has all that much benefit over just running rustc --cfg yourself...

I can understand the tacked on feeling of the feature as that was kind of the intention for the current implementation, i.e. re-use existing functionality and have a small footprint. I viewed the feature more as a shortcut currently missing from the `cargo rustc` interface that makes running `rustc --print cfg` yourself to obtain Cargo included compiler modifications cumbersome. 

I thought a `--cfg` or `--print-cfg` flag would (i) avoid any inspection and/or "magic" with extra arguments, (ii) be more consistent with other flags (`--unit-graph`) within the same subcommand, (iii) meet expectations in behavior, simplicity, and naming relative to the `rustc --print cfg` command, and (iv) gracefully handle future features like [multitargets](https://github.com/rust-lang/cargo/pull/8167). A feature for obtaining the Cargo-modified compiler configurations does _not_ need to be this implementation (see below).

>For example the current implementation just ignores everything else passed to `cargo rustc` effectively other than `--print-cfg`.

The current implementation follows a similar behavior to the `rustc --print cfg`, where the `--target` option is handled, but other flags and options appear to be ignored (I don't know for sure, I have not looked into the rustc interface too much). The `cargo rustc -Z unstable-options --target i686-pc-windows-gnu --print-cfg` command will print the host and the `i686-pc-windows-gnu` compiler configurations.

>I'm not really sure of the best way to solve and design this myself, however. We could maybe take a leaf from rustc's book and have something like `cargo print rustc-cfg` or maybe even `cargo config --print rustc-cfg` or I dunno. Something which is a general way of "print out Cargo's internal information" which we can extend in the future rather than "here's a very specific print flag on one command"

I am intrigued by a `cargo config` subcommand. It would be consistent with the `cargo metadata` subcommand and enable `--message-version`, `--mesage-format`, etc. flags as well. A `cargo print` command also might work, but then it would seem appropriate to have `cargo print --metadata` and `cargo print --config` (or equivalent). There might be flags and options specific to printing the configuration versus printing the metadata, which starts to feel messy.

The `cargo config` could print the "final" configuration obtained from environment variables and the TOML file hierarchy. A `rustc` field would be available in the default JSON output. The `rustc` field value would be something like the output in this current implementation. I don't think a `--rustc` (or equivalent) flag is even needed, other than to act like a filter and only output the compiler configuration versus other configurations, which feels unnecessary and something a downstream crate/subcommand should handle.

This does not cover CLI configurations, such as the multitarget feature or the `--target` option. Would the `cargo config` subcommand also need to replicate the `--target` option (and any other options/flags from other subcommands that modify the configuration)? I can see this more broad and extensible feature co-existing along with the narrow, compiler-specific feature that already accounts for CLI configurations.

If a `cargo config` (or equivalent) feature is ultimately the preferred implementation for obtaining Cargo's configuration and a `cargo rustc` specific implementation is not needed, then I am happy to drop/close this PR and start a new Issue/PR to work through the `cargo config` feature.

bors commented:
:umbrella: The latest upstream changes (presumably #8037) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#conflicts).

alexcrichton commented:
In general I mean learning about Cargo's configuration one way or another would be nice. For example `cargo config build.rustflags` would print out basically what you want here (sort of) for what to pass to rustc. 

I think `cargo rustc --print cfg` could make sense, even if it ignores all other flags, since as you say rustc  already does that. For this to work though I don't think Cargo should wrap everything in JSON since `cargo rustc` is typically a low-level layer around `rustc` itself, so we would presumably just invoke rustc with the various predefined flags we already have for the target.

Would that perhaps solve your issue? Implementing `cargo rustc --print foo` where that basically just forwards to `rustc --print foo` with the appropriately configured flags/etc that Cargo would otherwise be forwarding?

volks73 commented:
> In general I mean learning about Cargo's configuration one way or another would be nice. For example `cargo config build.rustflags` would print out basically what you want here (sort of) for what to pass to rustc.

This would be a nice feature, too. So, `cargo config` would function more like `git config`? Do you envision it also being able to "set" the values like `git config`?

> I think `cargo rustc --print cfg` could make sense, even if it ignores all other flags, since as you say rustc already does that. For this to work though I don't think Cargo should wrap everything in JSON since `cargo rustc` is typically a low-level layer around `rustc` itself, so we would presumably just invoke rustc with the various predefined flags we already have for the target.
> 
> Would that perhaps solve your issue? Implementing `cargo rustc --print foo` where that basically just forwards to `rustc --print foo` with the appropriately configured flags/etc that Cargo would otherwise be forwarding?

Yes, this works and makes perfect sense to me.

alexcrichton commented:
Yeah I've at least personally always wanted a `git config`-lookalike in Cargo. It probably wouldn't be easy to support writing config files just yet, but eventually yes I think we'd like to support that.

If `cargo rustc --print foo` works for you, though, perhaps this could switch to implementing that?

volks73 commented:
> If `cargo rustc --print foo` works for you, though, perhaps this could switch to implementing that?

Yes, works for me, and this can switch to implementing `cargo rustc --print foo`.

volks73 commented:
I have changed the implementation to be `cargo rustc --print cfg`, where `cfg` can be replaced with any of the accepted values for the `rustc --print <FOO>` command. The `cargo rustc --print <FOO>` command skips compiling, building a context, and bringing in the unit graph, but it does handle the `RUSTFLAGS` environment variable and similar fields defined in the Cargo configuration files hierarchy. This required changing the scope of some functions from private to public (crate). I am open to recommendations on alternatives.

Only the `--target <TRIPLE>` option from the `cargo rustc` command is recognized and "passed along" to rustc because several of the `--print <FOO>` options are compiler target-dependent, this enables support for multiple `--target` options, and the implementation re-uses the `env_args` function that uses the `--target` option value to search for configurations. Scanning through the extra arguments for a `--target` option felt a little clunky. If multiple `--target` options are present, then the output from each individual call with the `--target` option for rustc will be printed to STDOUT but separated by an empty line. Other than the `--target` option and any arguments after `--`, all other options for `cargo rustc` are ignored. Note, this means if `--target` is included in the extra arguments, an error will occur because `rustc` only allows one `--target` option.

The output from the `rustc --print <FOO>` command from the `cargo rustc --print <FOO>` command is simply dumped to STDOUT. There is no alternation of the output. An example invocation with just a single `--target` option that is not host:

```pwsh
PS C:\>cargo rustc -Z unstable-options --target i686-pc-windows-gnu --print cfg
debug_assertions
target_arch="x86"
target_endian="little"
target_env="gnu"
target_family="windows"
target_feature="fxsr"
target_feature="sse"
target_feature="sse2"
target_os="windows"
target_pointer_width="32"
target_vendor="pc"
windows

PS C:\>
```

An example output with multiple `--target` options:

```pwsh
PS C:\>cargo rustc -Z unstable-options -Z multitarget --target i686-pc-windows-gnu --target x86_64-unknown-linux-gnu --print cfg
debug_assertions
target_arch="x86"
target_endian="little"
target_env="gnu"
target_family="windows"
target_feature="fxsr"
target_feature="sse"
target_feature="sse2"
target_os="windows"
target_pointer_width="32"
target_vendor="pc"
windows

debug_assertions
target_arch="x86_64"
target_endian="little"
target_env="gnu"
target_family="unix"
target_feature="fxsr"
target_feature="sse"
target_feature="sse2"
target_os="linux"
target_pointer_width="64"
target_thread_local
target_vendor="unknown"
unix

PS C:\>
```

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -284,12 +288,76 @@ pub fn compile_ws<'a>(
+            .arg(print_opt_value)
+            .env_remove("RUSTC_LOG");
+        let (output, error) = rustc
+            .cached_output(&process)
```
I think it's ok to avoid `cached_output` here, that's just an optimization for normal builds. Here I think rustc should always be executed and we can just hook up stdout/stderr from rustc into Cargo's  own stdout/stderr.

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -284,12 +288,76 @@ pub fn compile_ws<'a>(
+    let mut kinds = build_config.requested_kinds.clone();
+    if kinds.is_empty() {
+        kinds.push(CompileKind::Host);
+    }
```
This seems like it's likely duplicated with somewhere else (handling an empty set of kinds). If so can it be deduplicated?

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -284,12 +288,76 @@ pub fn compile_ws<'a>(
+        process
+            .arg("-")
+            .arg("--crate-name")
+            .arg("___")
```
I don't think these three arguments are necessary?

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -284,12 +288,76 @@ pub fn compile_ws<'a>(
+        process
+            .arg("--print")
+            .arg(print_opt_value)
+            .env_remove("RUSTC_LOG");
```
This is ok to remove. If it should happen it should happen in `rustc.process()`, but otherwise I don't think there's any reason to single it out here.

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -284,12 +288,76 @@ pub fn compile_ws<'a>(
+        let (output, error) = rustc
+            .cached_output(&process)
+            .chain_err(|| "failed to run `rustc --print=<INFO>`")?;
+        if output.is_empty() {
```
I don't think there's any need to check this. We should just forward the request to rustc itself and rustc will do what it does .

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -284,12 +288,76 @@ pub fn compile_ws<'a>(
+        if print_empty_line {
+            writeln!(lock)?;
+        } else {
+            print_empty_line = true;
```
I think this could also be achieved with `.enumerate()` in this for-loop and by using `drop_println!` inside of Cargo at the head of the loop when the index is nonzero.

alexcrichton reviewed (commented):
Thanks! Could you also add a test for this?

volks73 commented:
> Could you also add a test for this?

Yes, I will create a test.

In the mean time, I believe I have addressed each of your comments with the latest round of commits. I tried using the `drop_println` macro, but I was getting a compile-time error about a `shell()` not existing. I did some poking around in the macro definition, but I became a little lost. I am not sure on the difference between the `drop_println` and the stdlib `println` macro, but the `println` macro appears to work without an issue.

volks73 commented:
Okay, now I see the difference between `println!` and `drop_println!`. I will need to investigate some more why I could not compile locally.

volks73 commented:
Turns out it was just importing the macro incorrectly. Either over or under thinking about the error message.

alexcrichton commented on src/cargo/ops/cargo_compile.rs:
```diff
@@ -289,12 +293,53 @@ pub fn compile_ws<'a>(
+            process.arg("--target").arg(t.short_name());
+        }
+        process.arg("--print").arg(print_opt_value);
+        let output = process.exec_with_output()?;
```
Instead of capturing the output here I think it's best to jsutt forward the output to Cargo's own stdout/stderr streams

volks73 replied:
I am sorry, but I am not sure how to forward the output to Cargo's own stdout/stderr streams. Do you have an example of doing this within the Cargo codebase for another process/cmd? I looked into the `ProcessBuilder` type and initially tried the `exec` method, but it just hanged. I believe the equivalent method in the `std::process::Command` type would be `status()` because it says it the child (rustc) inherients stdio from the parent (Cargo). Would this be the same as "forwarding"?

alexcrichton replied:
I believe you can basically call `exec` instead of `exec_with_output` here

alexcrichton commented on src/cargo/core/compiler/build_context/target_info.rs:
```diff
@@ -519,7 +519,7 @@ fn parse_crate_type(
 
 /// Helper for creating an error message when parsing rustc output fails.
-fn output_err_info(cmd: &ProcessBuilder, stdout: &str, stderr: &str) -> String {
+pub fn output_err_info(cmd: &ProcessBuilder, stdout: &str, stderr: &str) -> String {
```
I think this may no longer be needed

volks73 replied:
You are correct. It can be reverted back to private scope.

alexcrichton commented on src/cargo/core/compiler/build_context/target_info.rs:
```diff
@@ -552,7 +552,7 @@ fn output_err_info(cmd: &ProcessBuilder, stdout: &str, stderr: &str) -> String {
 /// Note that if a `target` is specified, no args will be passed to host code (plugins, build
 /// scripts, ...), even if it is the same as the target.
-fn env_args(
+pub fn env_args(
```
Is it perhaps possible to use `TargetInfo` directly rather than this low-level function? That structure handles things like `cfg`-related configuration as well.

volks73 replied:
I have made this change, but doing so does appear to slow down printing with multiple `--target` options. I think this is because `TargetInfo::new` does invoke multiple process calls to rustc and it is executing inside the loop for each compile kind/target. I am also guessing that these commands are not cached because a different target is being used on each loop. It is not a huge delay, but noticeable.

alexcrichton reviewed (commented):
Looking good!

volks73 commented:
I have changed to using the `exec` function instead of the `exec_with_output`. Something must have been up with my system because when I originally tried to use the `exec` function, I was getting no output and the command appeared to hang and not work. Now it works!

Anyways, I will on adding a test next.

volks73 commented:
I have added a couple of tests. These tests include using the `--print cfg` command with the `-Z multitarget` option, too.

alexcrichton commented:
Looks good to me, thanks!

@bors: r+

bors commented:
:pushpin: Commit 2a5355f9b3f2962286cd912245e9a3a0893c438f has been approved by `alexcrichton`

<!-- @bors r=alexcrichton 2a5355f9b3f2962286cd912245e9a3a0893c438f -->
<!-- homu: {"type":"Approved","sha":"2a5355f9b3f2962286cd912245e9a3a0893c438f","approver":"alexcrichton"} -->

bors commented:
:hourglass: Testing commit 2a5355f9b3f2962286cd912245e9a3a0893c438f with merge 8d8459a7f81bbb335b38665f980e03ce5c5b9ce9...
<!-- homu: {"type":"BuildStarted","head_sha":"2a5355f9b3f2962286cd912245e9a3a0893c438f","merge_sha":"8d8459a7f81bbb335b38665f980e03ce5c5b9ce9"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang/cargo/runs/1963035081)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang/cargo/runs/1963035081","builder_name":"checks-actions"} -->

ehuss commented:
@bors retry
macos: `The runner has received a shutdown signal.`

bors commented:
:hourglass: Testing commit 2a5355f9b3f2962286cd912245e9a3a0893c438f with merge 4ae4aadcb83924dfb69bc30a696f3edf5bdf06a3...
<!-- homu: {"type":"BuildStarted","head_sha":"2a5355f9b3f2962286cd912245e9a3a0893c438f","merge_sha":"4ae4aadcb83924dfb69bc30a696f3edf5bdf06a3"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/runs/1963628006)
Approved by: alexcrichton
Pushing 4ae4aadcb83924dfb69bc30a696f3edf5bdf06a3 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"alexcrichton","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/runs/1963628006"},"merge_sha":"4ae4aadcb83924dfb69bc30a696f3edf5bdf06a3"} -->
